pyvo/dal/adhoc.py

```diff
@@ -154,12 +168,68 @@ class DatalinkResultsMixin(AdhocServiceResultsMixin):
     """
     Mixing for datalink functionallity for results classes.
     """
+
     def iter_datalinks(self):
         """
         Iterates over all datalinks in a DALResult.
         """
-        for record in self:
-            yield record.getdatalink()
+        # To reduce the number of calls to the Datalink service, multiple
+        # IDs are sent in batches. The appropriate batch size is not available
+        # as each service has its own criteria for determining the maximum
+        # number of IDs processed per request. To overcome this limitation,
+        # this method initially sends all the available IDs and if the
+        # response is partial, uses the size of the response as the batch
+        # size.
+        if not hasattr(self, '_datalink'):
+            try:
+                self._datalink = self.get_adhocservice_by_ivoid(DATALINK_IVOID)
+            except DALServiceError:
+                self._datalink = None
+        if not hasattr(self, "_current_batch"):
+            # static vars
+            self._remaining_ids = []  # remaining IDs to processed
+            self._current_batch = None  # retrieved but not returned yet
+            self._current_ids = []  # retrieved but not returned
+            self._processed_ids = []  # retrived and returned IDs
+            self._batch_size = None  # size of the batch
+
+        for row in self:
+            if self._datalink:
+                if not self._current_ids:
+                    if self._batch_size is None:
+                        # first call.
+                        self.query = DatalinkQuery.from_resource(
+                            [_ for _ in self], self._datalink)
```

Comment:
I'm not happy about this parameter wrangling.
Maybe modify `from_resource` to have keyword parameters for different types of input?
The name is fine as the resource is always passed.

Reply:
How would you define the argument of `from_resouce`? Need to keep `row` for backwards compatibility and have an `additional_rows` one? I just cant think of a better way. Which I can rename `row` to `rows`. Please let me know if you have a better option - I'm open to it.

Reply:
`rows` sounds good indeed.

Reply:
Renamed to `rows`. Existing usages of the form `from_resouce(some_row, some_resource)`  would still work but not `from_resource(row=some_row, resource=some_resource)`. If we want to make that backwards compatible I would need to have defaults for both `rows` and `resource` and check for calls with `row` in `kwargs`. Too high of a price to pay I'm thinking, no?

Reply:
I don't think this should be a problem. At least i don't think anyone is using `from_resource` directly. Hmm.